chatbot-core/llm.py

```diff
@@ -0,0 +1,177 @@
+import logging
+from dataclasses import dataclass, field
+from typing import Any
+
+from embedding import (
+    OllamaEmbeddingModel,
+    OpenAIEmbeddingModel,
+    SentenceTransformerEmbeddingModel,
+)
+from langchain.chains.combine_documents import create_stuff_documents_chain
+from langchain.chains.history_aware_retriever import create_history_aware_retriever
+from langchain.chains.retrieval import create_retrieval_chain
+from langchain.memory import ConversationBufferWindowMemory
+from langchain.schema import AIMessage, HumanMessage
+from langchain_community.llms.ollama import Ollama
+from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
+from langchain_openai import ChatOpenAI
+from langchain_qdrant import QdrantVectorStore
+from qdrant_client import QdrantClient
+from utils import EmbeddingModelType
+
+from main.settings import (
+    EMBEDDING_MODEL_CHOICE,
+    OLLAMA_BASE_URL,
+    QDRANT_DB_COLLECTION_NAME,
+    QDRANT_DB_HOST,
+    QDRANT_DB_PORT,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class LLMBase:
+    """LLM Base containing common methods"""
+
+    qdrant_host: str = QDRANT_DB_HOST
+    qdrant_port: int = QDRANT_DB_PORT
+    qdrant_client: QdrantClient = field(init=False)
+    llm_model: Any = field(init=False)
+    memory: Any = field(init=False)
+
+    def __post_init__(self, mem_key: str = "chat_history", conversation_max_window: int = 3):
+        self.llm_model = None
+        self.qdrant_client = None
+        self.memory = None
+
+        self.qdrant_client = QdrantClient(host=self.qdrant_host, port=self.qdrant_port)
+        self.memory = ConversationBufferWindowMemory(k=conversation_max_window, memory_key=mem_key, return_messages=True)
+
+    def _system_prompt_for_retrieval(self):
+        """System prompt for information retrieval"""
+        return (
+            "Given a chat history and the latest user question "
+            "which might reference context in the chat history, "
+            "formulate a standalone question which can be understood "
+            "without the chat history. Do NOT answer the question, "
+            "just reformulate it if needed and otherwise return it as is."
+        )
+
+    def _system_prompt_for_response(self):
+        """
+        System prompt for response generation
+        """
+        system_prompt = """
+                You are an assistant for question-answering tasks.\n,
+                Use the following retrieved context to answer the question.\n,
+                If you don't get the answer from the provided context, \n,
+                say that 'I don't know. How can I help with the office related queries ?'
+                \n\n,
+                Context: {context}
+            """
+        return system_prompt
+
+    def get_selected_embedding_model(self, embedding_model_type: int = 1):
```

Comment:
Let's use enum here directly?
```suggestion
    def get_selected_embedding_model(self, embedding_model_type: EmbeddingModelType = EmbeddingModelType.SENTENCE_TRANSFORMES1):
```